### PR TITLE
Adding gofumpt

### PR DESCRIPTION
**Pull request by teivah**
Hey!
I was also interested in https://github.com/golangci/golangci-lint/issues/490 (adding gofumpt).
Here is a PR for it.

### CODE REVIEW

File: pkg/golinters/gofumpt.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+				}
+				output, err := format.Source(input, "")
+				if err != nil {
+					return nil, fmt.Errorf("error while running gofumpt: %w", err)
+				}
+				if !bytes.Equal(input, output) {
+					issues = append(issues, goanalysis.NewIssue(&result.Issue{
+						FromLinter: gofumptName,
+						Text:       "File is not `gofumpt`-ed",
+						Pos: token.Position{
```
Comment (tetafro) [inline pkg/golinters/gofumpt.go:56]: There is a `Replacement` parameter to automatically fix issues, which is very useful for gofumpt.

Comment (teivah) [reply pkg/golinters/gofumpt.go:56]: If I'm not mistaken, gofumpt provides only as a public function a way to get the gofumpt-ed version of the file (as a `[]byte`).
With `Replacement` you mean I should include the gofumpt-ed version of the whole file in the issue?

Comment (teivah) [reply pkg/golinters/gofumpt.go:56]: @tetafro Am I correct?

Comment (tetafro) [reply pkg/golinters/gofumpt.go:56]: > gofumpt provides only as a public function a way to get the gofumpt-ed version of the file

Yeah, sorry, I didn't think of it.

> With Replacement you mean I should include the gofumpt-ed version of the whole file in the issue?

I'm not sure if it's a right thing to do. Someone more experienced with golangci-lint architecture should answer.

---

Anyway, if there's no easy way to do this, just skip my first comment.

File: pkg/lint/lintersdb/manager.go [hunk 2]
```diff
@@ -191,6 +191,10 @@ func (m Manager) GetAllSupportedLinterConfigs() []*linter.Config {
 			WithPresets(linter.PresetFormatting).
 			WithAutoFix().
 			WithURL("https://golang.org/cmd/gofmt/"),
+		linter.NewConfig(golinters.NewGofumpt()).
+			WithPresets(linter.PresetFormatting).
+			WithAutoFix().
```
Comment (tetafro) [inline pkg/lint/lintersdb/manager.go:196]: If there is no autofix, you should remove this line

**Comment by tetafro** [comment]
Can you please add test to tests/testdata?

**Comment by teivah** [comment]
> Can you please add test to tests/testdata?

@tetafro I've just added a test. Just something to mention, previously I was setting the [error line](https://github.com/teivah/golangci-lint/blob/dffb7b4140ab4ee1d8d06c9e2fbfcd7fbb642e2c/pkg/golinters/gofumpt.go#L69) to 0. I wanted to change it to line 1 but in the test file, I can't do this:
```
// ERROR "File is not `gofumpt`-ed"
//args: -Egofumpt
package testdata
```
Hence, I set it to the last line number and it's working correctly.

**Comment by tetafro** [comment]
> Hence, I set it to the last line number and it's working correctly.

I've looked how gofmt linter is implemented. Seems like the right thing to do here is to re-use diff code from [gofmt linter](#https://github.com/golangci/golangci-lint/blob/master/pkg/golinters/gofmt_common.go#L50).

But I think using the last line is ok for the first iteration. Maybe just add `TODO` in the code.

**Comment by teivah** [comment]
> Thanks! This looks good to me.

@tetafro Any chance to get it merged then? :)

**Comment by tetafro** [comment]
> Any chance to get it merged then? :)

Yeah, sorry for delay

**Comment by teivah** [comment]
Thanks @tetafro

**Comment by gonzaloserrano** [comment]
Hi, how's the doc updated once a new linter is merged?

**Comment by teivah** [comment]
@tetafro Just out of curiosity, when do you plan roughly to make a new release?
Thanks

**Comment by tetafro** [comment]
@teivah this is a question for guys from core team. I just help with pr reviews from time to time.

**Pull request merged**